### PR TITLE
RocksDB Exploration II

### PR DESCRIPTION
**Pull request by johndebord**

### CODE REVIEW

File: test/chainbase_test.cpp [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+
+      time_t old_time{initial_time};
+      time_t new_time{initial_time};
+      
+      std::cout << "Filling initial database state...\n" << std::flush;
+      loggerman.print_progress(1,0);
+      for (size_t i{}; i < _gen_data.num_of_accounts_and_values()/10; ++i) {
+         new_time = time(NULL);
+         if (new_time != old_time) {
+            loggerman.log_tps      ({(new_time - initial_time), transactions_per_second/(new_time - initial_time)});
```
Comment (arhag) [inline test/chainbase_test.cpp]: It seems like you can just replace `transactions_per_second/(new_time - initial_time)` with 0 here.

File: test/chainbase_test.cpp [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+
+      time_t old_time{initial_time};
+      time_t new_time{initial_time};
+
+      std::cout << "Benchmarking...\n" << std::flush;
+      loggerman.print_progress(1,0);
+      for (size_t i{}; i < _gen_data.num_of_swaps(); ++i) {
+         new_time = time(NULL);
+         if (new_time != old_time) {
+            loggerman.log_tps      ({(new_time - initial_time), transactions_per_second/(new_time - initial_time)});
```
Comment (arhag) [inline test/chainbase_test.cpp]: Did you mean to use `initial_time` here or `old_time` when calculating the TPS? The current system is determining TPS by averaging over an ever expanding time window. If you want to calculate the TPS within certain narrow time windows, you could use `transactions_per_second/(new_time - old_time)` instead but then you would also need to reset `transactions_per_second` each time you log the TPS. And if you want TPS averaged over a rolling time window, you will need to do more work.

And once again same comment as above applies here about having better control over when this data is logged as time advances.

File: test/chainbase_test.cpp [hunk 3]
```diff
@@ -0,0 +1,506 @@
...
+            loggerman.log_ram_usage({(new_time - initial_time), _system_metrics.total_ram_currently_used()});
+            loggerman.log_cpu_load ({(new_time - initial_time), _system_metrics.calculate_cpu_load()});
+            loggerman.print_progress(i, _gen_data.num_of_accounts_and_values()/10);
+            old_time = new_time;
+         }
+
+         _database.start_undo_session(true);
+         
+         auto session{_database.start_undo_session(true)};
+         _database.start_undo_session(true);
```
Comment (arhag) [inline test/chainbase_test.cpp]: Why start so many undo sessions?

File: test/chainbase_test.cpp [hunk 4]
```diff
@@ -0,0 +1,506 @@
...
+      _bytes       = new uint8_t[ad._blob_size];
+      return *this;
+   }
+
+   bool operator< (const arbitrary_datum& rhs) const {
+      return std::string{*this} < std::string{rhs};
+   }
+
+   ~arbitrary_datum() {
+      delete[] _bytes;
```
[the remark this answers is not in the document]
Comment (johndebord) [reply test/chainbase_test.cpp]: `struct arbitrary_datum` has been removed; replaces with type alias to a vector or bytes.
https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L54-L60

File: test/chainbase_test.cpp [hunk 5]
```diff
@@ -0,0 +1,506 @@
...
+
+   arbitrary_datum& operator= (const arbitrary_datum& ad) {
+      _blob_size   = ad._blob_size;
+      _struct_size = ad._struct_size;
+      _bytes       = new uint8_t[ad._blob_size];
+      return *this;
+   }
+
+   bool operator< (const arbitrary_datum& rhs) const {
+      return std::string{*this} < std::string{rhs};
```
[the remark this answers is not in the document]
Comment (johndebord) [reply test/chainbase_test.cpp]: `struct arbitrary_datum` has been removed; replaced with type alias to `std::vector<uint8_t>`.

File: test/chainbase_test.cpp [hunk 6]
```diff
@@ -0,0 +1,506 @@
...
+   }
+
+   ~arbitrary_datum() {
+      delete[] _bytes;
+   }
+
+   operator std::string() const {
+      char* tmp{new char[_struct_size+1]};
+      memcpy(tmp, (void*)this, 16);
```
[the remark this answers is not in the document]
Comment (johndebord) [reply test/chainbase_test.cpp]: `struct arbitrary_datum` has been removed; replaced with type alias to `std::vector<uint8_t>`.

File: test/chainbase_test.cpp [hunk 7]
```diff
@@ -0,0 +1,506 @@
...
+   inline void _generate_values() {
+      std::cout << "Generating values...\n" << std::flush;
+      loggerman.print_progress(1,0);
+
+      time_t old_time{initial_time};
+      time_t new_time{initial_time};
+
+      for (size_t i{}; i < _num_of_accounts_and_values; ++i) {
+         // Add functionality to arbitrary_datum constructor to memset the blob by a certain value
+         _accounts.push_back(arbitrary_datum{(_uid(_dre) % _max_key_length+1),   _max_key_value});
```
[the remark this answers is not in the document]
Comment (johndebord) [reply test/chainbase_test.cpp]: https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L456
Resolved.

File: test/chainbase_test.cpp [hunk 8]
```diff
@@ -0,0 +1,506 @@
...
+public:
+   generated_data(size_t lower_bound_inclusive,
+                  size_t upper_bound_inclusive,
+                  size_t num_of_accounts_and_values,
+                  size_t num_of_swaps,
+                  size_t max_key_length,
+                  size_t max_key_value,
+                  size_t max_value_length,
+                  size_t max_value_value)
+      : _dre{static_cast<unsigned int>(time(0))}
```
[the remark this answers is not in the document]
Comment (johndebord) [reply test/chainbase_test.cpp]: https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L518-L520
https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L392
https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L414
The user now has the option to specify a unique seed.

File: test/chainbase_test.cpp [hunk 9]
```diff
@@ -0,0 +1,506 @@
...
+      time_t old_time{initial_time};
+      time_t new_time{initial_time};
+
+      for (size_t i{}; i < _num_of_accounts_and_values; ++i) {
+         // Add functionality to arbitrary_datum constructor to memset the blob by a certain value
+         _accounts.push_back(arbitrary_datum{(_uid(_dre) % _max_key_length+1),   _max_key_value});
+         _values.push_back  (arbitrary_datum{(_uid(_dre) % _max_value_length+1), _max_value_value});
+         
+         new_time = time(NULL);
+         if (new_time != old_time) {
```
[the remark this answers is not in the document]
Comment (johndebord) [reply test/chainbase_test.cpp]: https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L151-L233
https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L450
https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L459
https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L461
The time facilities have been updated to use `std::chrono`.

File: test/chainbase_test.cpp (hunk 1 above)
Comment (johndebord) [reply test/chainbase_test.cpp]: https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L598
Resolved.

File: test/chainbase_test.cpp (hunk 2 above)
Comment (johndebord) [reply test/chainbase_test.cpp]: https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L499-L503
https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L631-L651
https://github.com/EOSIO/chainbase/blob/chainrocks-tests/test/bench_tps.cpp#L678-L688
The user now has the option to specify how to measure the test.

File: test/chainbase_test.cpp (hunk 3 above)
Comment (johndebord) [reply test/chainbase_test.cpp]: My reasoning is to simulate an already long chain of `undo_session`s (blocks) for the test to operate on. In this case a new block is pushed for ever 10 accounts/balances created.

File: test/chainbase_test.cpp [hunk 10]
```diff
@@ -0,0 +1,506 @@
...
+
+private:
+   std::vector<std::pair<size_t,size_t>> _tps;
+   std::vector<std::pair<size_t,double>> _ram_usage;
+   std::vector<std::pair<size_t,double>> _cpu_load;
+   std::ofstream _data_file;
+};
+static logger loggerman;
+
+class system_metrics {
```
[the remark this answers is not in the document]
Comment (johndebord) [reply test/chainbase_test.cpp]: Planning to implement these.

File: test/chainbase_test.cpp [hunk 11]
```diff
@@ -0,0 +1,506 @@
...
+      return _uid(_dre);
+   }
+};
+
+struct account : public chainbase::object<0,account> {
+   template<typename Constructor, typename Allocator>
+   account(Constructor&& c, Allocator&& a) { c(*this); }
+   id_type id;
+   arbitrary_datum _account_key;
+   arbitrary_datum _account_value;
```
[the remark this answers is not in the document]
Comment (johndebord) [reply test/chainbase_test.cpp]: I think it might be useful to have an in-person discussion on this.

**Pull request closed without merging**